File: exercism/haskell/sum-of-multiples/src/SumOfMultiples.hs
module SumOfMultiples where

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples xs n = sum (filter isMultiple [1 .. n - 1])
  where
    isMultiple m = any (divides m) xs
    divides m = (0 ==) . (mod m)

sumOfMultiplesDefault :: Integer -> Integer
sumOfMultiplesDefault = sumOfMultiples [3, 5]

File: exercism/elixir/sum-of-multiples/sum_of_multiples.exs
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.

  The default factors are 3 and 5.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors \\ [3, 5]) do
    0..(limit-1) |> Enum.filter(&(is_multiple_of?(&1, factors)))
             |> Enum.sum
  end

  defp is_multiple_of?(n, factors) do
    Enum.any?(factors, &(rem(n, &1) == 0))
  end
end
